src/http_server.c

```diff
@@ -0,0 +1,219 @@
+	char *buffer = alloca(requiredSize);
+	snprintf(buffer, requiredSize, reply, strlen(userMessage), userMessage);
+
+	ssize_t write_result = write(socket, buffer, strlen(buffer));
+	(void)write_result;
+}
+
+static int process_request(int new_socket, char *id) {
+	log_info("Processing HTTP SAML request\n");
+
+	int flag = 1;
+	setsockopt(new_socket, IPPROTO_TCP, TCP_NODELAY, &flag, sizeof(int));
+
+	// Read the request
+	char request[1024];
+	ssize_t read_result = read(new_socket, request, sizeof(request) - 1 /*  Save one place for termination
+										in case the request is about to
+										fill the entire buffer. */
+				);
+
+	// Check for '=id' in the response
+	// If the recevied request from the server is larger than the buffer, the result will not be null-terminated.
+	// Causing strlen to behave wrong.
+	if (read_result < 0) {
+		log_error("Bad request\n");
+		send_status_response(new_socket, "Invalid redirect response from Fortinet server. VPN could not be established.");
+		return -1;
+	}
+
+	// Safety Null-terminate
+	request[sizeof(request) - 1] = 0;
+	request[read_result] = 0;
+
+	static const char *request_head = "GET /?id=";
+
+	if (strncmp(request, request_head, strlen(request_head)) != 0) {
+		log_error("Bad request\n");
+		send_status_response(new_socket, "Invalid redirect response from Fortinet server. VPN could not be established.");
+		return -1;
+	}
+
+	// Extract the id
+	static const char *token_delimiter = " &\r\n";
+	char *next_token = request + strlen(request_head); // strsep does modify the input argument and we don't want to loose our request pointer.
+	char *id_start = strsep(&next_token, token_delimiter);
+
+	if (next_token == NULL) {
+		// In case not found, next_token was set to NULL
+		// This should be invalid because we expect \r\n at the end of the GET request line
+		log_error("Bad request format\n");
+		send_status_response(new_socket, "Invalid formatting of Fortinet server redirect response. VPN could not be established.");
+		return -1;
+	}
+
+	// strsep inserted a NULL at the location of the delimiter.
+	int id_length = strlen(id_start);
+
+	if(id_length == 0 || id_length >= MAX_SAML_SESSION_ID_LENGTH) {
+		log_error("Bad request id\n");
+		send_status_response(new_socket, "Invalid SAML session id received from Fortinet server. VPN could not be established.");
+		return -1;
+	}
+
+	// It was checked above, that the length is smaller than MAX_SAML_SESSION_ID_LENGTH
+	strcpy(id, id_start);
+
+	for (int i = 0; i < id_length; i++) {
+		if (isalnum(id[i]) || id[i] == '-') continue;
+		log_error("Invalid id format\n");
+		send_status_response(new_socket, "Invalid SAML session id received from Fortinet server. VPN could not be established.");
+		return -1;
+	}
+
+	log_info("Extracted id: %s\n", id);
+	send_status_response(new_socket,
+			"SAML session id received from Fortinet server. VPN will be established..."
+			"\r\nYou may close this browser tab now.");
```

Comment:
I have found a nice improvement to the user experience: when a browser tab is opened from an external application, such as when user clicks on the printed link in their terminal, it is possible to close the tab using a single javascript command. The script won't have any effect if the user manually opened the tab and pasted the URL, but if they did that, they might as well close it manually.

I tested by hacking a separate function (shown below) for the success case. You may want to consider upgrading all the messages to HTML in order to have just one response function. In any case, the autoclose script should only be added to the success message.

```c
static void send_success_response(int socket) {
    static const char *reply = "HTTP/1.1 200 OK\r\n"
        "Content-Type: text/html\r\n"
        "Content-Length: %lu\r\n"
        "Connection: close\r\n\r\n"
        "%s";
    static const char *message =
        "<html>"
        "<p>"
        "SAML session id received from Fortinet server. "
        "VPN will be established..."
        "</p><p>"
        "You may close this browser tab now."
        "<p>"
        "<script>"
        "    window.close();"
        "</script>"
        "</html>";

	int requiredSize = snprintf(NULL, 0, reply, strlen(message), message) + 1;
	char *buffer = alloca(requiredSize);
	snprintf(buffer, requiredSize, reply, strlen(message), message);

	ssize_t write_result = write(socket, buffer, strlen(buffer));
	(void)write_result;
}
```

Reply:
Good that we are already down to the user experience concerns :D

I've added a little java script to the delivered message. Immediately closing the window is a bit rude because a user doesn't know what just happened, so I decided to have a 5 second timeout to read what is going on.